[WIP] Complete specs and guards for Kernel module and some submodules

eksperimental opened the pull request:
Cover the specs and add guards to functions for the Kernel module, and for some submodules,

- lib/elixir/lib/kernel/error_handler.ex
- lib/elixir/lib/kernel/lexical_tracker.ex
- lib/elixir/lib/kernel/typespec.ex
- lib/elixir/lib/kernel/utils.ex

I think for we don't need to do anything
- kernel/special_forms.ex

some submodules have been left untouched
- kernel/parallel_compiler.ex
- kernel/parallel_require.ex


There are a few specs that the compiler cannot understand, marked as:
`# TODO: FIX: following line gives error`

Once this is merged, I will update in a new PR the documentation based on the specs.
Specially the one with `no_return`.

I'm still unsure about the specs for some macros and that needs reviewing.

josevalim commented:
@eksperimental we should not introduce `no_return` in the majority of those cases. We only use `no_return` when there is a chance of error given correctly typed input. After all, almost all functions will given some input outside of its types.

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -107,15 +110,19 @@ defmodule Kernel.Utils do
   end
-  def raise(atom) when is_atom(atom) do
-    atom.exception([])
+
```
Please don't add those lines back if they are not related to the PR.

eksperimental replied:
sorry, i cleaned the PR of changes like this, i just missed this one

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -107,15 +110,19 @@ defmodule Kernel.Utils do
   @doc """
   Callback for raise.
   """
+  @spec raise(any) :: no_return
```
This is a good example of `no_return` btw. :)

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -94,9 +92,14 @@ defmodule Kernel.Utils do
   @doc """
   Announcing callback for defstruct.
   """
+  @spec announce_struct(mod) :: (:ok | {:struct_available, mod}) when mod: module
```
No need to spec this function. This is a callback for internal implementation details.

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -65,14 +69,8 @@ defmodule Kernel.Utils do
-        raise ArgumentError, "struct fields definition must be list, got: #{inspect other}"
-    end
-
+  @spec defstruct(module, Keyword.t) :: {struct, keys :: [atom], derive :: [module]}
```
No need to spec this function. This is a callback for internal implementation details.

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -19,7 +22,8 @@ defmodule Kernel.Utils do
   Callback for defdelegate.
   """
-  def defdelegate(fun, opts) do
+  @spec defdelegate(Macro.t, Keyword.t) :: [tuple]
```
No need to spec this function. This is a callback for internal implementation details.

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -6,8 +6,11 @@ defmodule Kernel.Utils do
   """
-  def destructure(list, count) when is_list(list), do: destructure_list(list, count)
-  def destructure(nil, count), do: destructure_nil(count)
+  @spec destructure(list | nil, non_neg_integer) :: list
```
No need to spec this function. This is a callback for internal implementation details.

josevalim commented on lib/elixir/lib/kernel/lexical_tracker.ex:
```diff
@@ -143,7 +144,7 @@ defmodule Kernel.LexicalTracker do
   end
 
-  def handle_cast({:alias_dispatch, module}, state) do
+  def handle_cast({:alias_dispatch, module}, state) when is_atom(module) do
```
Same thing for all other callbacks.

josevalim commented on lib/elixir/lib/kernel/error_handler.ex:
```diff
@@ -3,28 +3,36 @@
   end
 
+  @spec ensure_compiled(module, atom) :: boolean
+  @spec ensure_compiled(nil, any) :: false
```
I wouldn't add this clause because it doesn't add information for consumers of this API, at the documentation level (even though it is technically more correct).

eksperimental replied:
you mean `@spec ensure_compiled(nil, any) :: false` ?
or both specs?

josevalim replied:
Sorry, only the nil one. :)

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -2830,8 +2869,10 @@ defmodule Kernel do
+  @spec macro_exported?(module, atom, arity) :: boolean
+  def macro_exported?(module, macro, arity)
+      when :erlang.is_atom(module) and :erlang.is_atom(macro) and
+      :erlang.is_integer(arity) and (arity >= 0 and arity <= 255) do
```
Please indent it under `:erlang`.

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -2606,6 +2642,9 @@ defmodule Kernel do
   """
+  # TODO: FIX: following line gives error
+  # @spec (first .. last) :: Range.t(first, last) when first: integer, last: integer
+  @spec ((Macro.t | any)..(Macro.t | any)) :: Range.t | no_return
```
`..` in specs accept integer. I wouldn't spec this function because it is confusing to think about it receiving expressions.

eksperimental replied:
i was not very keen on this one either.

josevalim commented:
Thank you @eksperimental. I have added some comments. I think we should not spec callbacks nor macros. Specs for macros only have documentation purposes (since macros are called at compile time) and most times when we add a spec to a macro it may make things more confusing for users of the API.

eksperimental commented:
@josevalim could you please tell me the cases where given the types specified in the spec, the function will not raise. Cause they specs were created with that in mind.

josevalim commented:
@eksperimental I believe all of your `no_return` changes, except the one for Kernel.Utils.raise should not have a no_return. In doubt, you can check the Erlang equivalent: http://erlang.org/doc/man/erlang.html

eksperimental commented:
just to give an example
`binary_part` will throw no_return, given a particular situation.

even thought Erlang documentation says it only returns `binary()`, it is contradicting itself (IMO) 

> part(Subject, PosLen) -> binary()
>   Subject = binary()
>   PosLen = {Start :: integer() >= 0, Length :: integer()}
> If PosLen in any way references outside the binary, a badarg exception is raised.

it's just telling you it will return `no_return` if PosLen is out of range. 

same for the other ones. 
I was planning on adding the docs in a different PR, but I can create a second commit that shows when it will raise given the types provided in the specs.

josevalim commented:
> it's just telling you it will return no_return if PosLen is out of range.

That's what I meant. Not all conditions can be expressed with dialyzer but assuming valid inputs, it won't fail. Compare that to `raise` which was meant to fail.

eksperimental commented:
@josevalim I see what you mean. I thought a spec should mention when there would be a situation that fail given the provided input adhere to the specs.

I have addressed all the issues mentioned

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -957,7 +971,10 @@ defmodule Kernel do
+  @spec (nonempty_list ++ nonempty_list) :: nonempty_list
+  @spec (left ++ []) :: left when left: nonempty_list
+  @spec ([] ++ right) :: right when right: term
+  @spec (list ++ nonempty_improper_list(any, any)) :: nonempty_improper_list(any(), any())
```
I think this spec got considerably harder to understand what it does. And it is a spec that does not matter because the function is inlined anyway, so it only exists for documentation purposes. So I prefer to keep the previous one.

eksperimental replied:
Is there the Erlang documentation for `:erlang.++/2`? All i was able to find was able to find is for `:lists.append/2`, 

> append(List1, List2) -> List3
> Types:
> List1 = List2 = List3 = [T]
> T = term()

Maybe I could submit a patch to Erlang.

The other thing is the spec is complecated, but it describes exactly what the function does.
Maybe when I include the documentation it will be easier to follow.

josevalim replied:
The original one is still simpler, so I believe we should stick with it.

eksperimental replied:
@josevalim I realized I could simplify the specs from 4 to 2 clauses,

```elixir
  @spec ([] ++ right) :: right when right: var
  @spec (nonempty_list ++ maybe_improper_list(any, tail))
        :: nonempty_maybe_improper_list(any, tail) when tail: var
```

still the original one is simpler, but i like this one because it tells the user that if given an empty list on the left it will return whatever term on the right, making it clear it will not always return a list or an improper one as the docs currently express.

Once this PR is merged I will submit a PR improving the docs of many functions including this one.

so your call about this one and I think we are good to go.

josevalim replied:
Please let's keep the original one. Again, this exists mostly as a documentation purpose, and I think the current one is the clearest documentation wise.

eksperimental replied:
sure thing, no worries.

josevalim commented on lib/elixir/lib/kernel/lexical_tracker.ex:
```diff
@@ -147,7 +147,7 @@ defmodule Kernel.LexicalTracker do
     {:noreply, %{state | directives: add_dispatch(state.directives, module, :alias)}}
   end
 
-  def handle_cast({:add_import, module, fas, line, warn}, state) when is_atom(module) do
```
:+1:

josevalim commented on lib/elixir/lib/kernel/utils.ex:
```diff
@@ -110,12 +112,15 @@ defmodule Kernel.Utils do
   def raise(msg) when is_binary(msg) do
     RuntimeError.exception(msg)
   end
+
```
Let's not add those in this commit as you are not even changing those lines and `destructure` above does not introduce new lines either.

lexmag commented on lib/elixir/lib/kernel/error_handler.ex:
```diff
@@ -3,28 +3,35 @@
   end
 
-  def undefined_lambda(module, fun, args) do
+  @spec undefined_function(module, fun, list) :: term
```
`undefined_function` –> `undefined_lambda`

eksperimental replied:
good eye @lexmag :eye:

lexmag commented on lib/elixir/lib/kernel.ex:
```diff
@@ -2830,8 +2848,10 @@ defmodule Kernel do
-  def macro_exported?(module, macro, arity) do
+  @spec macro_exported?(module, atom, arity) :: boolean
+  def macro_exported?(module, macro, arity)
+      when :erlang.is_atom(module) and :erlang.is_atom(macro) and
```
We can omit `:erlang.` here.

fishcakez commented on lib/elixir/lib/kernel.ex:
```diff
@@ -780,8 +781,9 @@ defmodule Kernel do
+
   """
-  @spec tl(maybe_improper_list) :: maybe_improper_list
+  @spec tl(nonempty_maybe_improper_list) :: maybe_improper_list | term
```
Should use parametrised types here

eksperimental replied:
you mean: `nonempty_maybe_improper_list(any, any)` ?
if so, may I know what makes the difference.

fishcakez replied:
rather like `@spec tl(non_empty_maybe_improper_list(elem, tail) :: maybe_improper_list(elem, tail) | tail when elem: var, tail: var`. Using the variable means any inferred type information about the list itself remains. For example `tl([:a | :b, ...]) :: [:a | :b]`.

eksperimental replied:
this is great @fishcakez.
thank you so much!

fishcakez commented on lib/elixir/lib/kernel.ex:
```diff
@@ -559,7 +559,7 @@ defmodule Kernel do
   Allowed in guard tests. Inlined by the compiler.
   """
-  @spec node(pid | reference | port) :: node
+  @spec node(pid | reference | port) :: node | :nonode@nohost
```
`:nonode@nohost` is a node.

eksperimental replied:
I thought being explicit was good.
I have seen a few times other specs in the project where the return is `term | something_else`, `any | something_else` and `atom | nil`
here's a list:
https://gist.github.com/eksperimental/ba212be62d35b6e62d7876c3ce879e04

fishcakez commented on lib/elixir/lib/kernel.ex:
```diff
@@ -263,7 +263,7 @@ defmodule Kernel do
 
   """
-  @spec hd(maybe_improper_list) :: term
+  @spec hd(nonempty_maybe_improper_list) :: term
```
should use parametrised type here

fishcakez commented on lib/elixir/lib/kernel.ex:
```diff
@@ -881,7 +891,8 @@ defmodule Kernel do
   """
-  @spec (-number) :: number
+  @spec (-float) :: float
+  @spec (-integer) :: integer
```
`pos_integer` -> `neg_integer` etc?

eksperimental replied:
good idea. I will fix these

fishcakez commented on lib/elixir/lib/kernel.ex:
```diff
@@ -865,7 +875,7 @@ defmodule Kernel do
 
   """
-  @spec (+number) :: number
+  @spec (+arg) :: arg when arg: number
```
Not a fan of using `arg` as a variable

lexmag replied:
Agree, `arg` feels too generic.
I would actually prefer shadowing `(+number) :: number when number: number`. :bowtie:

eksperimental replied:
other alternative could be `value`

josevalim replied:
`(+number) :: number when number: number()` would work.

josevalim replied:
But I think value is better, otherwise it is too confusing.

eksperimental replied:
yes, let's not use `number`, that was the original intention

fishcakez commented on lib/elixir/lib/kernel.ex:
```diff
@@ -1146,6 +1164,7 @@ defmodule Kernel do
       false
 
   """
+  @spec (arg === arg) :: true when arg: term
```
This spec is incorrect, it saying for terms of same type, which are any term, the return is true. Then says same but false. I think previous spec is preferable. Overlapping definitions are awkward and dialyzer wont like them.

fishcakez commented on lib/elixir/lib/kernel/error_handler.ex:
```diff
@@ -3,28 +3,35 @@
-  def undefined_function(module, fun, args) do
+  @spec undefined_function(module, atom, list) :: term
+  def undefined_function(module, fun, args)
+      when is_atom(module) and is_atom(fun) and is_list(args) do
```
I dont think we require these guards because its not public api and the caller isnt going to send invalid arguments.

eksperimental replied:
does anybody else agree or disagree with this?

josevalim replied:
Yes, the undefined functions are callbacks.

eksperimental commented:
All issues addressed except these two:
- https://github.com/elixir-lang/elixir/pull/5391#discussion_r85908402
- https://github.com/elixir-lang/elixir/pull/5391#discussion_r85889434

whatyouhide commented on lib/elixir/lib/kernel.ex:
```diff
@@ -795,11 +798,13 @@ defmodule Kernel do
 
   """
-  @spec trunc(number) :: integer
+  @spec trunc(arg) :: arg when arg: integer
```
Is there a difference between this and saying `@spec trunc(number) :: integer` since it always returns an integer?

eksperimental replied:
so according to what you are saying that would be the same as writing: `@spec trunc(integer) :: integer` and making the whole "when..."  syntax unnecessary

eksperimental replied:
@whatyouhide 
from Typespecs in Erlang manual

> For example, the following specification defines the type of a polymorphic identity function:
>   -spec id(X) -> X.
> Notice that the above specification does not restrict the input and output type in any way. These types can be constrained by guard-like subtype constraints and provide bounded quantification:
> 
>   -spec id(X) -> X when X :: tuple().
> Currently, the :: constraint (read as «is a subtype of») is the only guard constraint that can be used in the 'when' part of a '-spec' attribute.
> 
> Note
> The above function specification uses multiple occurrences of the same type variable. That provides more type information than the following function specification, where the type variables are missing:
> 
>   -spec id(tuple()) -> tuple().
> The latter specification says that the function takes some tuple and returns some tuple. The specification with the X type variable specifies that the function takes a tuple and returns the same tuple.
>

whatyouhide replied:
Mh, I see. But to me it doesn't really make sense, for example, say you have a function that takes a function and returns the result of invoking the argument function, doubled:

```elixir
def call_then_double(fun) when is_function(fun, 0) do
  fun.() * 2
end
```

I would spec this as:

```elixir
@spec call_then_double((() -> result)) :: result when result: integer | float
```

but this doesn't mean that since the argument fun returns `result`, then the outer fun returns `result`, it rather means (at least in my head 😃) that if the argument function returns a value of type `result`  then `call_then_double/1` will return a value of the same type. 😕

eksperimental replied:
that's a good example, and I see what you mean.
Whether it works as you or I say, we should update the docs to explain this very clearly.

ericmj replied:
I think when they say "returns the same tuple" they really mean "returns tuple of the same type". So if you're given `{integer(), float()}` you are going to return the same type but values of the integer and float can be anything.

If they really mean the same exact value a lot of typespecs would not make sense. For example take a look at `lists:map/2` http://erlang.org/doc/man/lists.html#map-2. How would that make sense if all instances of `A` and `B` were the same value?

eksperimental replied:
sorry @ericmj but the example you gave uses no "when ..."

ericmj replied:
It does use when: https://github.com/erlang/otp/blob/maint/lib/stdlib/src/lists.erl#L1231-L1236

whatyouhide replied:
Yeah I think it's just a case of the funky docs™, I interpreted it same as @ericmj said. Then since this is the case, `@spec trunc(number) :: integer` is perfectly fine.

fishcakez replied:
The rule I use is that when a type is used as a parameter to another type I treat it as a set of types, and when the variable is never a parameter (i.e. only an argument and/or returned value) I treat it is as the exact value. For me this makes types spec as explicit as possible.

I am unaware of a situation where OTP breaks this rule but tools can interpret it as they wish. The lists example does not break this rule as `A` and `B` are used as parameters.

```elixir
@spec call_then_double((() -> result)) :: result when result: integer | float
```
Breaks all the rules I think because the returned value is neither the exact value nor the same set of integers returned by the fun. For example the fun may return `4..6` but then function returns `8..12`.

fishcakez replied:
The set of types returned by the anonymous function may not overlap with the set of types returned by `call_then_double/2`.

whatyouhide replied:
Mh, I feel like this behaviour is a bit confusing though :|

josevalim replied:
@eksperimental this is correct as is.

whatyouhide replied:
@josevalim do you mean "as is" as in how it is in the PR or how it is in the source?

josevalim replied:
In the PR. And :+1: to rename `arg` to `value`.

eksperimental commented on lib/elixir/lib/kernel.ex:
```diff
@@ -263,7 +263,7 @@ defmodule Kernel do
 
   """
-  @spec hd(maybe_improper_list) :: term
+  @spec hd(nonempty_maybe_improper_list(elem, term)) :: elem when elem: var
```
I would prefer to stick to one, since we have `var`, `term` and `any`. is there a reason when to use `var` over the other ones?

fishcakez commented:
yeah

On 2 November 2016 at 23:19, Eksperimental <notifications@github.com> wrote:

> *@eksperimental* commented on this pull request.
> ------------------------------
>
> In lib/elixir/lib/kernel.ex
> <https://github.com/elixir-lang/elixir/pull/5391>:
>
> > @@ -263,7 +263,7 @@ defmodule Kernel do
>        1
>
>    """
> -  @spec hd(maybe_improper_list) :: term
> +  @spec hd(nonempty_maybe_improper_list) :: term
>
> @fishcakez <https://github.com/fishcakez> is it fine like this:?
> @spec hd(nonempty_maybe_improper_list(elem, term)) :: elem when elem: term
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/5391>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6JTa0-prFYvIgn17bnX3_HGcDnk9Xdks5q6RqHgaJpZM4KlaUq>
> .
>

eksperimental commented:
all issues have been addressed, except this one, 
https://github.com/elixir-lang/elixir/pull/5391#discussion_r86242507
so what is the verdict on that?

josevalim commented on lib/elixir/lib/kernel/typespec.ex:
```diff
@@ -325,6 +335,7 @@ defmodule Kernel.Typespec do
   ## Macro callbacks
 
   @doc false
+  @spec defspec(atom, Macro.t, Macro.Env.t) :: Keyword.t
```
This is a callback, you can remove this one. :)

eksperimental replied:
removed

eksperimental commented on lib/elixir/lib/kernel.ex:
```diff
@@ -603,7 +603,8 @@ defmodule Kernel do
   """
-  @spec round(number) :: integer
+  @spec round(float) :: integer
+  @spec round(arg) :: arg when arg: integer
```
based on the previous discussion, do you guys want me to rename `arg` to `value` ?

eksperimental replied:
fixed

eksperimental commented:
latest changes in the last commit. no more issues left :sweat_smile:

josevalim commented on lib/elixir/lib/kernel.ex:
```diff
@@ -2830,9 +2847,12 @@ defmodule Kernel do
-    function_exported?(module, :__info__, 1) and :lists.member({macro, arity}, module.__info__(:macros))
+  @spec macro_exported?(module, atom, arity) :: boolean
+  def macro_exported?(module, macro, arity)
+      when is_atom(module) and is_atom(macro) and is_integer(arity) and
```
Beautiful indentation. ;)

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
